resource/name.go

```diff
@@ -107,6 +128,18 @@ func RemoveRemoteName(n Name) Name {
 	return tempName
 }
 
+// AddPartID returns a new name with part ID added. This will replace an existing part ID.
+func (n Name) AddPartID(partID string) Name {
+	tempName := NewName(n.API, n.ShortName())
+	tempName.MachinePartID = partID
+	return tempName
+}
+
+// RemovePartID returns a new name with part ID removed.
+func (n Name) RemovePartID() Name {
+	return NewName(n.API, n.ShortName())
+}
+
```

Comment:
[q] do we have places were we might want to lookup or compare using either a name with OR without the part id?

If not, I think it would be helpful to have two different structs - one with the machine id and one without - e.g. `Name`, which includes the machine id, and `NameWithoutMachineID`, which does not (having the reverse is fine too - can probably be private/internal). This will prevent use from accidentally use a `With/Without` comparison when we meant to use the other one - it's also pretty easy to convert to the version we need.

--------

[minor] Somewhat unrelated from my above comment - I slightly prefer we name these functions `With/Without` instead of `Add/Remove`, just to clarify that nothing is being mutated. We do document that behavior so no big deal if you prefer to leave as is.

```suggestion
// WithPartID returns a new name with part ID added. This will replace an existing part ID.
func (n Name) WithPartID(partID string) Name {
	tempName := NewName(n.API, n.ShortName())
	tempName.MachinePartID = partID
	return tempName
}

// WithoutPartID returns a new name with part ID removed.
func (n Name) WithoutPartID() Name {
	return NewName(n.API, n.ShortName())
}
```

Reply:
I'll play around with an internal struct without machine id for usage around comparisons

Reply:
playing around with it, I don't think we should make a separate struct.

A user should never use the struct without machine id, so this would be purely internal usage for internal structures that are tracking resources. If so, I would want to make it private, but that creates an import cycle if I want to embed NameWithoutMachineID back in Name. We could have resource.Name not embed NameWithoutMachineID, but that makes everything harder to maintain. Ultimately we just have to be vigilant around adding tests.

Reply:
what about adding a "public" struct in the `resource` package but putting it in the `internal` folder so it can't be imported by external clients of rdk? 

e.g. 

```go
// internal/resource/resource.go

package "resource"

type NameWithoutMachineID struct {
// ...
}
```

With this setup, we can import `resource.NameWithoutMachineID` wherever we need to, but external users won't have access to the this struct.

Reply:
I guess this still has the downside of requiring a separate import e.g. `go.viam.com/rdk/internal/resource`. Anyway, I'm not gonna block on this - my general view is that we should use the static type system to make our lives easier whenever possible - so long as it's reasonable to achieve.

Thanks for exploring this.

Reply:
> my general view is that we should use the static type system to make our lives easier whenever possible - so long as it's reasonable to achieve.

ya, I agree with that, I do feel like this one is a bit difficult to achieve and ends up complicating the types a bit too much